go/client.go

```diff
@@ -141,31 +154,71 @@ func (s *ClientState) SendClientStarted() error {
 
 // NewMaster modifies the client state by setting the current master to a new
 // one.
-func (s *ClientState) NewMaster(server string) {
+func (s *RedisSystem) NewMaster(server string) {
 	logInfo("setting new master: %s", server)
 	s.currentMaster = NewRedisShim(server)
 }
 
-// DetermineInitialMaster tries to read the current master from disk.
-func (s *ClientState) DetermineInitialMaster() {
-	if !MasterFileExists(s.GetConfig().RedisMasterFile) {
+// UpdateMasterFile writes the known masters information to the redis master file.
+func (s *ClientState) UpdateMasterFile() {
+	path := s.GetConfig().RedisMasterFile
+	systems := make(map[string]string, 0)
+	for system, rs := range s.redisSystems {
+		if rs.currentMaster == nil {
+			systems[system] = ""
+		} else {
+			systems[system] = rs.currentMaster.server
+		}
+	}
+	content := MarshalMasterFileContent(systems)
+	WriteRedisMasterFile(path, content)
+}
+
+// DetermineInitialMasters tries to read the current masters from disk
+// and establish the system name to redis shim mapping.
+func (s *ClientState) DetermineInitialMasters() {
+	path := s.GetConfig().RedisMasterFile
+	if !MasterFileExists(path) {
+		s.UpdateMasterFile()
 		return
 	}
-	server := ReadRedisMasterFile(s.GetConfig().RedisMasterFile)
-	if server != "" {
-		s.NewMaster(server)
+	masters := RedisMastersFromMasterFile(path)
+	invalidSystems := make([]string, 0)
+	for system, server := range masters {
+		rs := s.EnsureKnownSystem(system)
+		if server != "" {
+			rs.NewMaster(server)
+		}
+		if rs.currentMaster == nil || !rs.currentMaster.IsMaster() {
+			invalidSystems = append(invalidSystems, system)
+			rs.currentMaster = nil
+		}
+	}
+	if len(invalidSystems) > 0 {
+		logInfo("clearing systems from master file %s", invalidSystems)
+		s.UpdateMasterFile()
 	}
 }
 
+func (s *ClientState) EnsureKnownSystem(system string) *RedisSystem {
+	rs := s.redisSystems[system]
+	if rs == nil {
+		rs = &RedisSystem{system: system}
+		s.redisSystems[system] = rs
```

Comment:
If I understand correctly, we trust all incoming system name strings here and add them to the map. 
This could lead to a DoS if a client generated random system strings e.g. in ping messages, right?

From the function name, I would assume we would only want to allow known system names.

Reply:
The client has no way of knowing which systems are known. This information is only available from the config server. And the client needs to trust the server.

Reply:
I could rename the method to something like RegisterSystem.

Reply:
Doh. I failed to see that this function is only used in the client code. Thanks for clarifying.
Renaming it to `RegisterSystem` sounds good though.